src/core/meshcop/dataset.hpp

```diff
@@ -51,16 +51,16 @@ namespace MeshCoP {
  * This class represents MeshCop Dataset.
  *
  */
-class Dataset
+class Dataset : private otOperationalDatasetTlvs
```

Comment:
I have a suggestion here (to resolve the casting issue). 
- Let's not have `Dataset` inherit from `otOperationalDatasetTlvs`
- Instead we can add method to convert from one type to the other.
   - A `Dataset::ConvertTo(otOperationalDatasetTlv &aDatasetTlvs) const;`
   - And maybe a constructor `Dataset::Dataset(Type aType, otOperationalDatasetTlv &aDatasetTlvs)` (or an `Init`).

We will have extra copying but looking at  `DatasetLocal::Read(otOperationalDatasetTlvs &aDataset` we already do need copying.

-------
Alternate (maybe longer term) is to change `Dataset` class to not track the `mType` and `mUpdateTime` and have maybe `DatasetManager` or `DatsetLocal` do it?

Reply:
I was trying to avoid the copy, but that would require more work as you pointed out. I reverted to the suggested simpler approach. Please review.